File: homework1/pla.m
function [w times] = pla(x, y, ratio, method)
rs = size(x, 1);
cs = size(x, 2);
X = [ones(rs, 1) x];
w = zeros(1, cs + 1);
times = 0;
while any((Y = sign(X * w')) != y)
    if (strcmp(method, 'order'))
        r = min(find(Y != y));
    elseif (strcmp(method, 'random'))
        r = randsample(find(Y != y), 1);
    end
    w = w + ratio * y(r) * X(r, :);
    ++times;
end

File: homework1/pla.R
pla <- function(x, y, ratio, method) {
  rs <- nrow(x)
  cs <- ncol(x)
  X <- cbind(as.matrix(rep(1, rs)), x)
  w <- as.matrix(rep(0, cs + 1))
  times <- 0
  while (any((Y <- sign(X %*% w)) != y)) {
    index <- which(Y != y)
    if (method == "order") {
      r = min(index)
    } else {
      r = sample(index, size = 1)
    }
    w <- w + ratio * y[r] * t(h(X, r))
    times <- times + 1
  }
  return(list(w, times))
}
